internal/services/elastic/elastic.go

```diff
@@ -0,0 +1,257 @@
+	}
+	es8Client, err := elasticsearch.NewTypedClient(esConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create elastic client: %w", err)
+	}
+	eClient := Service{
+		esClient: es8Client,
+		log:      logger,
+		settings: settings,
+	}
+	return &eClient, nil
+}
+
+// ESClient returns the underlying elastic client used by the service.
+//
+// This function exists to maintain previous behavior for endpoint who are not yet using the elastic service.
+func (s *Service) ESClient() *elasticsearch.TypedClient {
+	return s.esClient
+}
+
+// GetHistoryParams defines the parameters for retrieving history from elastic.
+type GetHistoryParams struct {
+	// StartTime is the start time of the history.
+	StartTime time.Time
+	// EndTime is the end time of the history.
+	EndTime time.Time
+	// DeviceID is the ID of the device.
+	DeviceID string
+	// PrivilegeIDs is the list of privilege IDs that the user has.
+	PrivilegeIDs []privileges.Privilege
+	// Buckets is the number of time intervals to divide the history into.
+	Buckets int
+}
+
+// GetHistory retrieves the history of a device from elastic. The history is divided into buckets and one data point is selected from each bucket.
+// The data points are selected by the first data point in the bucket that has a vehicle field.
+// The result is a list of data points.
+func (s *Service) GetHistory(ctx context.Context, params GetHistoryParams) ([]json.RawMessage, error) {
+	if params.Buckets < 1 {
+		return nil, fmt.Errorf("non-positive number of buckets: %w", ErrInvalidParams)
+	}
+	// build the query
+	req := buildHistoryQuery(params)
+
+	// perform the request
+	res, err := s.esClient.Search().Index(s.settings.DeviceDataIndexNameV2).Request(&req).Perform(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to complete request: %w", err)
+	}
+
+	//nolint:errcheck // we don't care about the error closing the body.
+	defer res.Body.Close()
+
+	// check for errors any status code over 299 is an error
+	if res.StatusCode >= http.StatusMultipleChoices {
+		body, _ := io.ReadAll(res.Body) // we ignore the error here because we are already in an error state
+		return nil, fmt.Errorf("got status code %d from Elastic message = %s", res.StatusCode, string(body))
+	}
+
+	body, err := io.ReadAll(res.Body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read response body: %w", err)
+	}
+
+	// depending on the start and end time we may get len(results) <=  buckets + 1
+	// pass the max number of values to getHitsFromHistoryResponse to ensure we only get the desired number of values
+	return getHitsFromHistoryResponse(body, params.Buckets), nil
+}
+
+// SetDefaultHistoryParams sets the default values for the history params.
+// If the buckets are less than or equal to 0, it will be set to 1000.
+// If the end time is zero, it will be set to the current time.
+// If the start time is zero, it will be set to 2 weeks from the end time.
+func (g *GetHistoryParams) SetDefaultHistoryParams() {
```

Comment:
Oh, we are doing this. Could we put it next to the struct?

Reply:
Yeah